system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+ *    Settings_create() again, or by restarting the application.
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *    Success or negated failure code
+ *
+ ****************************************************************************/
+
+int settings_clear(void)
+{
+  int ret;
+
+  ret = pthread_mutex_lock(&g_settings.mtx);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  if (!g_settings.initialized)
+    {
+      ret = -EACCES;
+      goto errout;
+    }
+
+  memset(map, 0, sizeof(map));
+  g_settings.hash = 0;
+
+  save();
+
+errout:
+  pthread_mutex_unlock(&g_settings.mtx);
+
+  while (g_settings.wrpend)
+    {
+      usleep(10 * 1000); /* Sleep for 10ms */
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: settings_create
+ *
+ * Description:
+ *    Creates a new setting.
+ *
+ *    If the setting is found to exist in any of the storages, it will
+ *    be loaded. Else, it will be created and the default value will be
+ *    assigned.
+ *
+ * Input Parameters:
+ *        key         - the key of the setting.
+ *        type        - the type of the setting.
+ *        ...         - the default value of the setting.
+ *
+ * Returned Value:
+ *    Success or negated failure code
+ *
+ ****************************************************************************/
+
+int settings_create(FAR char *key, enum settings_type_e type, ...)
+{
+  int ret = OK;
+  FAR setting_t *setting = NULL;
+  int i;
+
+  DEBUGASSERT(type != SETTING_EMPTY);
+
+  DEBUGASSERT(strlen(key));
+  if (strlen(key) == 0)
+    {
+      return -EINVAL;
+    }
+
+  DEBUGASSERT(strlen(key) < CONFIG_SYSTEM_SETTINGS_KEY_SIZE);
+  if (strlen(key) >= CONFIG_SYSTEM_SETTINGS_KEY_SIZE)
+    {
+      return -EINVAL;
+    }
+
+  DEBUGASSERT(isalpha(key[0]) && (sanity_check(key) == OK));
+  if (!isalpha(key[0]) || (sanity_check(key) < 0))
+    {
+      return -EINVAL;
+    }
+
+  ret = pthread_mutex_lock(&g_settings.mtx);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  if (!g_settings.initialized)
+    {
+      ret = -EACCES;
+      goto errout;
+    }
+
+  for (i = 0; i < CONFIG_SYSTEM_SETTINGS_MAP_SIZE; i++)
+    {
+      if (strcmp(key, map[i].key) == 0)
+        {
+          setting = &map[i];
+
+          /* We found a setting with this key name */
+
+          goto errout;
+        }
+
+      if (map[i].type == SETTING_EMPTY)
+        {
+          setting = &map[i];
+
+          /* This setting is empty/unused - we can use it */
+
+         break;
+        }
+    }
+
+  DEBUGASSERT(setting);
+
+  if (setting && (setting->type != type))
+    {
+      strncpy(setting->key, key, CONFIG_SYSTEM_SETTINGS_KEY_SIZE);
+      setting->key[CONFIG_SYSTEM_SETTINGS_KEY_SIZE - 1] = '\0';
+      setting->type = type;
+
+      va_list ap;
+      va_start(ap, type);
+
+      switch (type)
+      {
+        case SETTING_STRING:
+          {
+            FAR char *str = va_arg(ap, FAR char *);
```

Comment:
I can't see any difference?

Reply:
Check here: https://github.com/fjpanag/code_for_nuttx/blob/00b20ed3cc70b5872fcb2063ec4d3a45247f3b79/settings/settings.c#L266

It will only re-create the setting if it is of an incompatible type.
As it is now it will re-init it in case it is of different type, even a compatible one!

Reply:
As above - my stupidity with git cloning